lib/utils/cobs.c

```diff
@@ -1,104 +1,638 @@
+/* Verify sufficient data available for block */
+static inline int verify_block_data(const struct net_buf *src, uint8_t offset)
+{
+	if (src->len < (offset - 1)) {
+		return -EINVAL;
+	}
+	return 0;
+}
 
-		if (end_delim != delimiter) {
+/* Copy data bytes from encoded block */
+static inline int copy_block_data_bytes(struct net_buf *src, struct net_buf *dst,
+					uint8_t count, const uint8_t delimiter)
+{
+	for (uint8_t i = 0; i < count; i++) {
+		uint8_t byte = net_buf_pull_u8(src);
+
+		if (byte == delimiter) {
 			return -EINVAL;
 		}
+		net_buf_add_u8(dst, byte);
+	}
+	
+	return 0;
+}
+
+/* Insert delimiter between blocks if needed */
+static inline void insert_delimiter_if_needed(struct net_buf *dst, uint8_t offset,
+					      const struct net_buf *src,
+					      const uint8_t delimiter)
+{
+	/* If this wasn't a maximum offset and we have more data,
+	 * there was a delimiter here in the original data
+	 */
+	if (offset != 0xFF && src->len > 0) {
+		net_buf_add_u8(dst, delimiter);
+	}
+}
+
+int cobs_decode(struct net_buf *src, struct net_buf *dst, uint32_t flags)
+{
+	uint8_t delimiter = COBS_FLAG_CUSTOM_DELIMITER(flags);
+
+	int ret = validate_trailing_delimiter(src, flags, delimiter);
+	if (ret != 0) {
+		return ret;
 	}
 
 	while (src->len > 0) {
 		/* Pull the COBS offset byte */
 		uint8_t offset = net_buf_pull_u8(src);
 
-		if (offset == delimiter && !(flags & COBS_FLAG_TRAILING_DELIMITER)) {
-			return -EINVAL;
+		ret = validate_offset(offset, delimiter, flags);
+		if (ret != 0) {
+			return ret;
 		}
 
-		/* Verify we have enough data */
-		if (src->len < (offset - 1)) {
-			return -EINVAL;
+		ret = verify_block_data(src, offset);
+		if (ret != 0) {
+			return ret;
 		}
 
 		/* Copy offset-1 bytes */
-		for (uint8_t i = 0; i < offset - 1; i++) {
-			uint8_t byte = net_buf_pull_u8(src);
+		ret = copy_block_data_bytes(src, dst, offset - 1, delimiter);
+		if (ret != 0) {
+			return ret;
+		}
+
+		insert_delimiter_if_needed(dst, offset, src, delimiter);
+	}
+
+	return 0;
+}
+
+/* ========================================================================
+ * Streaming Encoder Implementation
+ * ======================================================================== */
+
+void cobs_encode_init(struct cobs_encode_state *self)
+{
+	if (!self) {
+		return;
+	}
+
+	self->src_frag = NULL;
+	self->src_offset = 0;
+	self->block_code = 0;
+	self->block_pos = 0;
+}
+
+void cobs_encode_reset(struct cobs_encode_state *self)
+{
+	if (!self) {
+		return;
+	}
+
+	self->src_frag = NULL;
+	self->src_offset = 0;
+	self->block_code = 0;
+	self->block_pos = 0;
+}
+
+/* ========================================================================
+ * Source Navigation Helpers
+ * ======================================================================== */
+
+/* Skip empty fragments to find next available data */
+static inline void skip_empty_fragments(struct cobs_encode_state *self)
+{
+	while (self->src_frag && self->src_offset >= self->src_frag->len) {
```

Comment:
No implicit boolean conversions (same for others)

```suggestion
	while (self->src_frag != NULL && self->src_offset >= self->src_frag->len) {
```